src/librbd/ImageWatcher.cc

```diff
@@ -0,0 +1,643 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+#include "librbd/ImageWatcher.h"
+#include "librbd/AioCompletion.h"
+#include "librbd/ImageCtx.h"
+#include "cls/lock/cls_lock_client.h"
+#include "cls/lock/cls_lock_types.h"
+#include "include/encoding.h"
+#include "include/stringify.h"
+#include "common/errno.h"
+#include <sstream>
+#include <boost/bind.hpp>
+#include <boost/function.hpp>
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::ImageWatcher: "
+
+namespace librbd {
+
+static const std::string WATCHER_LOCK_TAG = "internal";
+static const std::string WATCHER_LOCK_COOKIE_PREFIX = "auto";
+
+static const uint64_t	NOTIFY_TIMEOUT = 5000;
+static const uint8_t	NOTIFY_VERSION = 1;
+
+enum {
+  NOTIFY_OP_ACQUIRED_LOCK = 0,
+  NOTIFY_OP_RELEASED_LOCK = 1,
+  NOTIFY_OP_REQUEST_LOCK  = 2,
+  NOTIFY_OP_HEADER_UPDATE = 3
+};
+
+class FunctionContext : public Context {
+public:
+  FunctionContext(const boost::function<void()> &callback)
+    : m_callback(callback)
+  {
+  }
+
+  virtual void finish(int r) {
+    m_callback();
+  }
+private:
+  boost::function<void()> m_callback;
+};
+
+
+ImageWatcher::ImageWatcher(ImageCtx &image_ctx)
+  : m_image_ctx(image_ctx), m_watch_ctx(*this), m_handle(0),
+    m_lock_owner_state(LOCK_OWNER_STATE_NOT_LOCKED),
+    m_finisher(new Finisher(image_ctx.cct)),
+    m_watch_lock("librbd::ImageWatcher::m_watch_lock"), m_watch_error(0),
+    m_aio_request_lock("librbd::ImageWatcher::m_aio_request_lock"),
+    m_retrying_aio_requests(false)
+{
+  m_finisher->start();
+}
+
+ImageWatcher::~ImageWatcher()
+{
+  m_finisher->stop();
+  delete m_finisher;
+}
+
+bool ImageWatcher::is_lock_supported() const {
+  return ((m_image_ctx.features & RBD_FEATURE_EXCLUSIVE_LOCK) != 0 &&
```

Comment:
D'oh ... never mind.  I thought you were talking about the "owner lock" concept instead of the actual "owner_lock" in ImageCtx.  Same issue w/ my response to your other comment.